### PR TITLE
Refactor resolve request

### PR DESCRIPTION
**Pull request by notaphplover**
## Description

- `_resolveRequest` function has been refactored to extract a `_resolveBinding` function.
- Added generic `T` type to `resolveInstance` call at _resolveRequest.
- Renamed Binding T generic to TActivated.
- Updated `Binding.factory` to be `FactoryCreator<TActivated> | null`.

## Related Issue

#1321, #1322

## Motivation and Context

This change reduces the code complexity of the _resolveRequest funcion and also adds more concrete types

## How Has This Been Tested?

- Code compiles
- All test passes successfully

## Types of changes

- [x] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the changelog.

### CODE REVIEW

**Comment by PodaruDragos** [comment]
Nice work, LGTM

sâm., 1 mai 2021, 12:19 notaphplover ***@***.***> a scris:

**Comment by notaphplover** [comment]
There's a test failing on node 14 (es6), I'll have a look but I'll be missing for some hours. @tonyhallett could you have a look? 

EDIT: if it actually sounds rude, I'm just asking a favor :smiley: (english is not my first language, sry)

File: src/interfaces/interfaces.ts [hunk 1]
```diff
@@ -61,8 +61,8 @@ namespace interfaces {
         dynamicValue: DynamicValue<TActivated> | null;
         scope: BindingScope;
         type: BindingType;
-        implementationType: Newable<TActivated> | null;
-        factory: FactoryCreator<TActivated> | null;
+        implementationType: Newable<TActivated> | TActivated | null;
+        factory: FactoryCreator<any> | null;
```
Comment (PodaruDragos) [inline src/interfaces/interfaces.ts]: i'm just curios. can we not use `unknown` here ?

Comment (notaphplover) [reply src/interfaces/interfaces.ts]: I guess it would force adding some castings

**Review by notaphplover — COMMENTED** [review]
Great! I think it's possible to reduce the complexity of those functions, I can go for it this night

File: src/resolution/resolver.ts [hunk 2]
```diff
@@ -51,103 +51,128 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+    } else if (binding.type === BindingTypeEnum.Function) {
+        result = binding.cache!;
+        binding.activated = true;
+    } else if (binding.type === BindingTypeEnum.Constructor) {
+        result = binding.implementationType as T;
+    } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
+        result = invokeFactory(
+            "toDynamicValue",
+            binding.serviceIdentifier,
+            () => (binding.dynamicValue!)(request.parentContext)
```
Comment (notaphplover) [inline src/resolution/resolver.ts]: It's easy not to notice about this notation (`!`) I often avoid it in favor of the casting, but it's just my opinion

File: src/resolution/resolver.ts [hunk 3]
```diff
@@ -51,103 +51,128 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+            binding.implementationType as interfaces.Newable<T>,
+            childRequests,
+            _resolveRequest<T>(requestScope)
+        );
+    } else {
+        // The user probably created a binding but didn't finish it
+        // e.g. container.bind<T>("Something"); missing BindingToSyntax
+        const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
+        throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
+    }
```
Comment (notaphplover) [inline src/resolution/resolver.ts]: Maybe we could replace this for a `switch` control structure.

```ts

switch (binding.type) {
    case BindingTypeEnum.DynamicValue:
        result = _getREsolvedDynamicValueBinding(...)
        break;
}

...

function _getREsolvedDynamicValueBinding(...) {
    if (binding.dynamicValue) {
        return invokeFactory(
            "toDynamicValue",
            binding.serviceIdentifier,
            () => (binding.dynamicValue!)(request.parentContext)
        );
    } else {
         return undefined;
    }
}

```

We should throw an error if result is `undefined` after this switch:

```ts
    if (result === undefined) {
        // The user probably created a binding but didn't finish it
        // e.g. container.bind<T>("Something"); missing BindingToSyntax
        const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
        throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
    }
```

File: src/resolution/resolver.ts [hunk 4]
```diff
@@ -51,103 +51,146 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+        _isRequestScope(binding) &&
+        requestScope !== null &&
+        requestScope.has(binding.id)
+    ) {
+        return requestScope.get(binding.id);
+    }
+    return null;
+}
+
+const _isSingletonScope = (binding:interfaces.Binding<any>): boolean => {
```
Comment (notaphplover) [inline src/resolution/resolver.ts]: We prefer `unknown` over `any`
```suggestion
const _isSingletonScope = (binding:interfaces.Binding<unknown>): boolean => {
```

File: src/resolution/resolver.ts [hunk 5]
```diff
@@ -51,103 +51,146 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+    return binding.scope === BindingScopeEnum.Singleton;
+}
 
-            if (isPromise(result)) {
-                result = result.catch((ex) => {
-                    // allow binding to retry in future
-                    binding.cache = null;
-                    binding.activated = false;
+const _isRequestScope = (binding:interfaces.Binding<any>): boolean => {
```
Comment (notaphplover) [inline src/resolution/resolver.ts]: We prefer `unknown` over `any`
```suggestion
const _isRequestScope = (binding:interfaces.Binding<unknown>): boolean => {
```

Comment (tonyhallett) [reply src/resolution/resolver.ts]: Done

File: src/resolution/resolver.ts (hunk 4 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: done

File: src/resolution/resolver.ts (hunk 2 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: Exclamation marks stand out to me ! 
Feel free to put back !

Comment (tonyhallett) [reply src/resolution/resolver.ts]: I am refactoring and they should go

File: src/resolution/resolver.ts (hunk 3 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: I will get there.......

**Comment by tonyhallett** [comment]
@notaphplover @PodaruDragos  Ok to squash and merge ?

File: src/resolution/resolver.ts (hunk 3 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: Done

File: src/resolution/resolver.ts (hunk 2 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: Gone !

**Comment by tonyhallett** [comment]
done making changes

**Comment by notaphplover** [comment]
> done making changes

I'm back at home, on my way to review it :)

**Comment by notaphplover** [comment]
Ok I have some changes to request, I already have them unstaged on my local repo so if you agree I can easily push them

File: src/resolution/resolver.ts [hunk 6]
```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
-        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
-            result = invokeFactory(
-                "toProvider",
-                binding.serviceIdentifier,
-                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
-            result = resolveInstance(
+            break;
+        case "Constructor":
```
Comment (notaphplover) [inline src/resolution/resolver.ts]: We prefer `BindingTypeEnum`. I know it's not a real enum, we will iterate though this

File: src/scope/scope.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import { BindingScopeEnum, interfaces } from "../inversify";
+import { isPromise } from "../utils/async";
+
+export const _tryGetFromScope = <T>(
```
Comment (notaphplover) [inline src/scope/scope.ts]: An export function is public so the underscore symbol should be removed

File: src/scope/scope.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      binding.scope === BindingScopeEnum.Request &&
+      requestScope !== null &&
+      requestScope.has(binding.id)
+  ) {
+      return requestScope.get(binding.id);
+  }
+  return null;
+}
+
+export const _saveToScope = <T>(
```
Comment (notaphplover) [inline src/scope/scope.ts]: Same here :)

File: src/utils/binding_utils.ts [hunk 9]
```diff
@@ -1,6 +1,54 @@
...
+            boundValue = binding.factory;
+            break;
+    }
+    if (boundValue === null) {
+            // The user probably created a binding but didn't finish it
+        // e.g. container.bind<T>("Something"); missing BindingToSyntax
+        const serviceIdentifierAsString = getServiceIdentifierAsString(binding.serviceIdentifier);
+        throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifierAsString}`);
+    }
+}
```
Comment (notaphplover) [inline src/utils/binding_utils.ts]: Brilliant!

File: src/utils/binding_utils.ts [hunk 10]
```diff
@@ -1,6 +1,54 @@
...
+import * as ERROR_MSGS from "../constants/error_msgs";
 export const multiBindToService = (container: interfaces.Container) =>
     (service: interfaces.ServiceIdentifier<any>) =>
         (...types: interfaces.ServiceIdentifier<any>[]) =>
             types.forEach((t) => container.bind(t).toService(service));
+
+export const _ensureFullyBound = (binding:interfaces.Binding<unknown>):void => {
+    let boundValue:unknown = null;
+    switch(binding.type){
+        case "ConstantValue":
```
Comment (notaphplover) [inline src/utils/binding_utils.ts]: We prefer `BindingTypeEnum`

File: src/utils/binding_utils.ts [hunk 11]
```diff
@@ -1,6 +1,54 @@
...
+            break;
+    }
+    if (boundValue === null) {
+            // The user probably created a binding but didn't finish it
+        // e.g. container.bind<T>("Something"); missing BindingToSyntax
+        const serviceIdentifierAsString = getServiceIdentifierAsString(binding.serviceIdentifier);
+        throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifierAsString}`);
+    }
+}
+export type FactoryType = "toDynamicValue" | "toFactory" | "toAutoFactory" | "toProvider";
```
Comment (notaphplover) [inline src/utils/binding_utils.ts]: We can use a real enum:

```ts

export enum FactoryType {
    AutoFactory = "toAutoFactory",
    DynamicValue = "toDynamicValue",
    Factory = "toFactory",
    Provider = "toProvider",
}

```

File: src/utils/binding_utils.ts [hunk 12]
```diff
@@ -1,6 +1,54 @@
...
+    switch(binding.type){
+        case "Factory":
+            return { factory: binding.factory,factoryType:"toFactory"};
+        case "Provider":
+            return { factory: binding.provider,factoryType:"toProvider"};
+        case "DynamicValue":
+            return { factory: binding.dynamicValue,factoryType:"toDynamicValue"};
+    }
+    return null;
+}
```
Comment (notaphplover) [inline src/utils/binding_utils.ts]: Nice, but this function returns null as unexpected value. I think its cleaner to throw an exception:

```ts

export const _getFactoryDetails = (binding: interfaces.Binding<unknown>): FactoryDetails => {
    switch(binding.type){
        case BindingTypeEnum.Factory:
            return { factory: binding.factory, factoryType: FactoryType.Factory };
        case BindingTypeEnum.Provider:
            return { factory: binding.provider, factoryType: FactoryType.Provider };
        case BindingTypeEnum.DynamicValue:
            return { factory: binding.dynamicValue, factoryType: FactoryType.DynamicValue };
        default:
            throw new Error(`Unexpected factory type ${binding.type}`);
    }
}

```

This way you don't need to cast `_getFactoryDetails(binding) as FactoryDetails` later

**Review by notaphplover — COMMENTED** [review]
These are the changes :)

File: src/resolution/resolver.ts (hunk 6 above)
Comment (tonyhallett) [reply src/resolution/resolver.ts]: Ok

File: src/scope/scope.ts (hunk 7 above)
Comment (tonyhallett) [reply src/scope/scope.ts]: Agreed. I was on an underscore roll.

File: src/scope/scope.ts (hunk 8 above)
Comment (tonyhallett) [reply src/scope/scope.ts]: Agreed

File: src/utils/binding_utils.ts (hunk 9 above)
Comment (tonyhallett) [reply src/utils/binding_utils.ts]: Thanks

File: src/utils/binding_utils.ts (hunk 10 above)
Comment (tonyhallett) [reply src/utils/binding_utils.ts]: Ok

File: src/utils/binding_utils.ts (hunk 11 above)
Comment (tonyhallett) [reply src/utils/binding_utils.ts]: Sure.  It was copy paste of existing code.
You might want to remove toAutoFactory from it.

File: src/utils/binding_utils.ts (hunk 12 above)
Comment (tonyhallett) [reply src/utils/binding_utils.ts]: Agreed.

**Comment by tonyhallett** [comment]
> Ok I have some changes to request, I already have them unstaged on my local repo so if you agree I can easily push them

Please do !

File: src/resolution/resolver.ts [hunk 13]
```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+    return result;
+}
 
+const _resolveBinding = <T>(
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>,
+): T | Promise<T> => {
+    return _resolveInScope(requestScope,binding, () => {
+        let result = _getResolvedFromBinding<T>(requestScope, request, binding);
```
Comment (tonyhallett) [inline src/resolution/resolver.ts]: No need to supply the generic
`_getResolvedFromBinding<T>(requestScope, request, binding);`

File: src/resolution/resolver.ts [hunk 14]
```diff
@@ -51,103 +31,86 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
...
+    requestScope: interfaces.RequestScope,
+    request: interfaces.Request,
+    binding:interfaces.Binding<T>): T | Promise<T> => {
+    let result: T | Promise<T> | undefined;
+    const childRequests = request.childRequests;
+    _ensureFullyBound(binding);
+    switch(binding.type){
+        case "ConstantValue":
+        case "Function":
+            result = binding.cache as T | Promise<T>;
```
Comment (tonyhallett) [inline src/resolution/resolver.ts:71]: I don't think that this should be here.
`binding.activated = true;`
This is related to singleton scope not a resolved value from the binding type.  After it is resolved from binding the singleton scope sets binding.activated to true.

**Review by tonyhallett — COMMENTED** [review]
Only other thing that could be done is to further replace T with TActivated.  I personally don't mind if we do this at a later stage.

File: src/resolution/resolver.ts (hunk 14 above)
Comment (notaphplover) [reply src/resolution/resolver.ts:71]: We can do it in this PR, I created an issue to solve this so I think it's ok to change it

Comment (tonyhallett) [reply src/resolution/resolver.ts:71]: Great !

Comment (tonyhallett) [reply src/resolution/resolver.ts:71]: Ok for you to change these ?

Comment (tonyhallett) [reply src/resolution/resolver.ts:71]: One other thing that could be changed here given we are refactoring "resolve"

NextArgs can be generic as well as GetArgs.
Then _getAllArgs / _getNotAllArgs and _planAndResolve can be changed.

Comment (notaphplover) [reply src/resolution/resolver.ts:71]: Nice! I removed `bindingActivated` assignment. Can you update `NextArgs` and `GetArgs`?

Comment (tonyhallett) [reply src/resolution/resolver.ts:71]: Will do be about 30 minutes

**Review by notaphplover — COMMENTED** [review]
LGTM! I think we can squash and merge once the changelog has the BCs!

**Comment by tonyhallett** [comment]
Ok can I leave that with you ?

**Comment by notaphplover** [comment]
> Ok can I leave that with you ?

I said to add BC to the changelog because I read https://github.com/inversify/InversifyJS/issues/1319, but im not sure now. I think the cache refactor task is the right one

**Comment by notaphplover** [comment]
So, I think we can squash and merge now, Ill do it this afternoon, once Im at home :)

**Pull request merged**